scalatest.dotty/src/main/scala/org/scalatest/DiagrammedExprMacro.scala

```diff
@@ -38,112 +38,128 @@ object DiagrammedExprMacro {
-    import refl._
-    
     def apply(l: Expr[_], name: String, r: List[Expr[_]]): Expr[T] =
       Select.overloaded(l.unseal, name, Nil, r.map(_.unseal)).seal.cast[T]
 
+    def selectField(o: Expr[_], name: String): Expr[T] = Select.unique(o.unseal, name).seal.cast[T]
+
+    def default: Expr[DiagrammedExpr[T]] = '{ DiagrammedExpr.simpleExpr($expr, ${ getAnchor(expr) } ) }
+
     expr.unseal.underlyingArgument match {
-      case Apply(Select(lhs, op), rhs :: Nil) =>
+      case Apply(Select(New(_), _), _) => default
+
+      case IsApply(apply) if isXmlSugar(apply) => default
+
+      case IsApply(apply) if isJavaStatic(apply) => default
+
+      case Select(This(_), _) => default
+
+      case IsSelect(x) if x.symbol.flags.is(Flags.Object) => default
+
+      case IsSelect(x) if isJavaStatic(x) => default
+
+      case sel @ Select(qual, name) =>
+        type S
+        implicit val tpS: quoted.Type[S] = qual.tpe.seal.asInstanceOf[quoted.Type[S]]
+        val obj = parse[S](qual.seal.asInstanceOf[Expr[S]])
+        val anchor = getAnchorForSelect(refl)(sel.asInstanceOf[Select])
+
+        '{
+          val o = $obj
+          DiagrammedExpr.selectExpr(o, ${ selectField('{o.value}, name) }, $anchor)
+        }
+
+      case Block(stats, expr) =>
+        Block(stats, parse(expr.seal.cast[T]).unseal).seal.cast[DiagrammedExpr[T]] // call parse recursively using the expr argument if it is a block
+
+      case Apply(sel @ Select(lhs, op), rhs :: Nil) =>
+        val anchor = getAnchorForSelect(refl)(sel.asInstanceOf[Select])
         op match {
           case "||" | "|" =>
-            val left = parse(lhs.seal.cast[T & Boolean])
-            val right = parse(rhs.seal.cast[T & Boolean])
+            val left = parse(lhs.seal.cast[Boolean & T])
+            val right = parse(rhs.seal.cast[Boolean & T])
+
             '{
               val l = $left
               val r = $right
               if (l.value) l
-              else DiagrammedExpr.applyExpr(l, r :: Nil, r.value, ${ getAnchor(expr) })
+              else DiagrammedExpr.applyExpr(l, r :: Nil, r.value, $anchor)
             }
           case "&&" | "&" =>
-            val left = parse(lhs.seal.cast[T & Boolean])
-            val right = parse(rhs.seal.cast[T & Boolean])
+            val left = parse(lhs.seal.cast[Boolean & T])
+            val right = parse(rhs.seal.cast[Boolean & T])
             '{
               val l = $left
               val r = $right
-              if (l.value) DiagrammedExpr.applyExpr(l, r :: Nil, r.value, ${ getAnchor(expr) })
+              if (l.value) DiagrammedExpr.applyExpr(l, r :: Nil, r.value, $anchor)
               else l
             }
           case _ =>
-            val left = parse(lhs.seal.cast[Any])
-            val right = parse(rhs.seal.cast[Any])
+            type S
+            implicit val tpS: quoted.Type[S] = lhs.tpe.seal.asInstanceOf[quoted.Type[S]]
+            val left = parse[S](lhs.seal.asInstanceOf[Expr[S]])
```

Comment:
@nicolasstucki I have to use a few hacks like the above. I'm not sure if there is better ways to achieve the goal.

Reply:
That was the goal of the old `Sealed` abstraction that we never merged. I will revisit it.